File: src/write-cptrelease.go
/*
	This code is used by CPT to write patching related data to file "/etc/cptrelease".
	These are the content of file
	{
	"caseNumber": "012345",
	"patchingOwner": "sbhathej",
	"patchingStartTime": "02-26-2018 18:58:48",
	"patchingEndTime": "02-26-2018 18:59:02",
	"currentPatchBundle": "2018.01",
	"lastPatchBundle": "2017.11"
	}
*/

package main

import (
	"bytes"
	"encoding/json"
	"flag"
	"fmt"
	"io/ioutil"
	"log"
	"os"
	"os/exec"
	"strings"
)

const lastBundle = `/bin/grep VERSION /etc/sfdc-release | awk -F"-" '{print $3"."$4}'`

// Struct to create Json file
type cpt struct {
	CaseNumber         string `json:"caseNumber"`
	PatchingOwner      string `json:"patchingOwner"`
	PatchingstartTime  string `json:"patchingStartTime"`
	PatchingendTime    string `json:"patchingEndTime"`
	CurrentPatchBundle string `json:"currentPatchBundle"`
	LastPatchBundle    string `json:"lastPatchBundle"`
}

// Function to execute UNIX commands
func RunCommands(cmd string) (string, error) {
	cmdOut := exec.Command("bash", "-c", cmd)
	var stdout, stderr bytes.Buffer
	cmdOut.Stdout = &stdout
	cmdOut.Stderr = &stderr
	err := cmdOut.Run()
	if err != nil {
		err = fmt.Errorf("command %s failed with %s\n", cmd, err)
		return "", err
	}

	outStr, errStr := string(stdout.Bytes()), string(stderr.Bytes())
	if errStr != "" {
		err = fmt.Errorf("cmd.Run() failed with %s\n", errStr)
		return "", err
	}
	return outStr, nil
}

// Check sudo user
func CheckSudoUser() error {
	var user = os.Getenv("USER")
	if user != "root" {
		err := fmt.Errorf("error: sadly you need to run me as root")
		return err
	}
	return nil
}

// Function main
func main() {

	var cptrelease cpt

	err := CheckSudoUser()
	if err != nil {
		log.Fatal(err)

	}

	// Command line arguments
	bundle := flag.String("b", "", "Current OS Patchset")
	caseNum := flag.String("c", "", "GUS Change CaseNumber")
	last := flag.Bool("last", false, "Check last bundle")
	help := flag.Bool("h", false, "Command line help")

	flag.Parse()

	if *help {
		fmt.Println(`usage: write-cptrelease -b <current Bundle> -c <case_number>`)
		fmt.Println(`usage: write-cptrelease -last`)
		os.Exit(0)

	}

	// Get Timestamp
	timeStamp, err := RunCommands(`date '+%m-%d-%Y %H:%M:%S'`)
	if err != nil {
		log.Print(err)
	}
	timeStamp = strings.Trim(timeStamp, "\n")

	// To check if "/etc/cptrelease" file exists
	raw, err := ioutil.ReadFile("/etc/cptrelease")
	if err != nil {
		log.Printf("can't get the last patch bundle, %s", err)
	} else {
		err := json.Unmarshal(raw, &cptrelease)

		if err != nil {
			os.Exit(1)
		}
	}

	if *last {
		lastPatchBundle, err := RunCommands(lastBundle)

		if err != nil {
			log.Printf("%s", err)
		}

		cptrelease.LastPatchBundle = strings.Trim(lastPatchBundle, "\n")
		cptrelease.PatchingstartTime = string(timeStamp)
	} else {
		cptrelease = cpt{CaseNumber: *caseNum, PatchingOwner: os.Getenv("SUDO_USER"), PatchingendTime: string(timeStamp),
			CurrentPatchBundle: *bundle, LastPatchBundle: cptrelease.LastPatchBundle, PatchingstartTime: cptrelease.PatchingstartTime}
	}

	jsonData, err := json.MarshalIndent(cptrelease, "", "\t")

	f, err := os.Create("/etc/cptrelease")
	if err != nil {
		log.Fatal(err)
	}
	f.WriteString(string(jsonData))

	// use os.Chmod() function.
	err = os.Chmod("/etc/cptrelease", 0644)
	if err != nil {
		log.Println(err)
	}
	defer f.Close()

}

File: bin/write_cptrelease.py
#!/usr/bin/env python

# imports
from argparse import ArgumentParser
import json
from datetime import datetime
from os import environ, chmod
import logging

CPTFILE = '/etc/cptrelease'
SFDCFILE = '/etc/sfdc-release'


def read_jsonfile(filename):
    try:
        with open(filename) as f:
            f_read = f.read()
            return json.loads(f_read)
    except (IOError, ValueError) as e:
        logger.warning(e)
        return {}


def check_sudouser():
    if environ['USER'] != "root":
        logger.error("Sadly you need to run me as root")
        exit(1)


def custom_logger():
    """
    This function is used to set custom logger.
    :return:
    """
    # Setting up custom logging
    c_logger = logging.getLogger('write_cptrelease.py')

    c_logger.setLevel(logging.DEBUG if args.verbose is True else logging.INFO)

    # create a Stream handler
    ch = logging.StreamHandler()
    ch.setLevel(logging.DEBUG if args.verbose is True else logging.INFO)

    # create a logging format
    formatter = logging.Formatter('%(asctime)s - %(name)s - %(levelname)s - %(message)s')
    ch.setFormatter(formatter)

    # add the handlers to the logger
    if not c_logger.handlers:
        c_logger.addHandler(ch)
    c_logger.propagate = False

    return c_logger


if __name__ == "__main__":
    parser = ArgumentParser(prog='write_cptrelease.py', usage='\n%(prog)s --last \n%(prog)s -b 2018.01 -c 012345 ')
    parser.add_argument("--last", "-l", dest="lastbundle", action="store_true", default=False,
                        help="To get last patch Bundle")
    parser.add_argument("-b", dest="bundle", help="Bundle Version")
    parser.add_argument("-c", dest="casenumber", help="Case Number")
    parser.add_argument("--verbose", "-v", action="store_true", dest="verbose", default=False, help="verbosity")

    args = parser.parse_args()
    logger = custom_logger()

    check_sudouser()

    json_data = read_jsonfile(CPTFILE)

    if args.lastbundle:
        try:
            with open(SFDCFILE) as f:
                f_content = f.readlines()
            last_bundle = ".".join(f_content[0].split()[1].split('-')[2:])  # Hack to get last installed OS bundle
            json_data['lastPatchBundle'] = last_bundle
        except IOError as e:
            logger.warning("Can not open file %s", e)
            json_data['lastPatchBundle'] = ""

        json_data['patchingStartTime'] = datetime.now().strftime("%m-%d-%Y %H:%M:%S")
        json_data['patchingOwner'] = environ['SUDO_USER']

    if args.bundle and args.casenumber:
        json_data['currentPatchBundle'] = args.bundle
        json_data['patchingEndTime'] = datetime.now().strftime("%m-%d-%Y %H:%M:%S")
        json_data['caseNumber'] = args.casenumber

    try:
        with open(CPTFILE, 'w') as f:
            json.dump(json_data, f)
    except IOError as e:
        logger.error(e)

    chmod(CPTFILE, 0644)
